core/src/services/huggingface/backend.rs

```diff
@@ -0,0 +1,313 @@
+// under the License.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use http::StatusCode;
+use log::debug;
+
+use super::core::HuggingFaceCore;
+use super::error::parse_error;
+use super::lister::HuggingFaceLister;
+use super::message::HuggingFaceStatus;
+use crate::raw::*;
+use crate::*;
+
+/// [HuggingFace](https://huggingface.co/docs/huggingface_hub/package_reference/hf_api)'s API support.
+#[doc = include_str!("docs.md")]
+#[derive(Default, Clone)]
+pub struct HuggingFaceBuilder {
+    repo_type: Option<String>,
+    repo_id: Option<String>,
+    revision: Option<String>,
+    root: Option<String>,
+    read_token: Option<String>,
+}
+
+impl Debug for HuggingFaceBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Builder");
+
+        ds.field("repo_type", &self.repo_type);
+        ds.field("repo_id", &self.repo_id);
+        ds.field("revision", &self.revision);
+        ds.field("root", &self.root);
+
+        if self.read_token.is_some() {
+            ds.field("read_token", &"<redacted>");
+        }
+
+        ds.finish()
+    }
+}
+
+impl HuggingFaceBuilder {
+    /// Set repo type of this backend. Default is model.
+    ///
+    /// Available values:
+    /// - model
+    /// - dataset
+    ///
+    /// Currently, only models and datasets are supported.
+    /// [Reference](https://huggingface.co/docs/hub/repositories)
+    pub fn repo_type(&mut self, repo_type: &str) -> &mut Self {
+        if !repo_type.is_empty() {
+            self.repo_type = Some(repo_type.to_string());
+        }
+        self
+    }
+
+    /// Set repo id of this backend. This is required.
+    ///
+    /// Repo id consists of the account name and the repository name.
+    ///
+    /// For example, model's repo id looks like:
+    /// - meta-llama/Llama-2-7b
+    ///
+    /// Dataset's repo id looks like:
+    /// - databricks/databricks-dolly-15k
+    pub fn repo_id(&mut self, repo_id: &str) -> &mut Self {
+        if !repo_id.is_empty() {
+            self.repo_id = Some(repo_id.to_string());
+        }
+        self
+    }
+
+    /// Set revision of this backend. Default is main.
+    ///
+    /// Revision can be a branch name or a commit hash.
+    ///
+    /// For example, revision can be:
+    /// - main
+    /// - 1d0c4eb
+    pub fn revision(&mut self, revision: &str) -> &mut Self {
+        if !revision.is_empty() {
+            self.revision = Some(revision.to_string());
+        }
+        self
+    }
+
+    /// Set root of this backend.
+    ///
+    /// All operations will happen under this root.
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if !root.is_empty() {
+            self.root = Some(root.to_string());
+        }
+        self
+    }
+
+    /// Set the read token of this backend.
+    ///
+    /// This is optional.
```

Comment:
There are write tokens and read tokens.
https://huggingface.co/docs/hub/security-tokens#how-to-manage-user-access-tokens
However, the write token only allows to write in the repo with write permissions (owner's repo or repo of owner's organization).

Reply:
> There are write tokens and read tokens. https://huggingface.co/docs/hub/security-tokens#how-to-manage-user-access-tokens However, the write token only allows to write in the repo with write permissions (owner's repo or repo of owner's organization).

Yes, as @Inokinoki mentioned, HuggingFace only has either read or write access bound with a token. I initially added both, but since we only support read access, I dropped the write_token. I'll add a comment for a clearer explanation.

Reply:
I feel like we don't need to separate the read and write token. We don't support write operations now and we don't have a plan to do so. We can add them in the future.

Reply:
I would also name it `token` to be consistent with the existing huggingface ecosystem (not only the HF libraries but also third-party libraries integrated with the Hub). At the moment, tokens are mostly `read` or `write` it is not set in stone. We recently introduced in beta tokens that have other types of permissions (`api-inference` for instance). So I wouldn't differentiate token types here and name everything `token`.